File: frontend/src/types/filingExtraction.ts
export type FieldStatus = "confirmed" | "suggested" | "missing" | "rejected";

export interface FieldEvidence {
  source_type?: string | null;
  page_from?: number | null;
  page_to?: number | null;
  chunk_id?: string | null;
  text?: string | null;
  validation_notes?: string | null;
}

export interface FieldCandidate {
  value: string;
  normalized_value?: string | null;
  confidence: number;
  status: FieldStatus;
  evidence?: FieldEvidence | null;
}

export interface FieldResult {
  field_key: string;
  field_label: string;
  status: FieldStatus;
  value?: string | null;
  normalized_value?: string | null;
  confidence: number;
  evidence?: FieldEvidence | null;
  suggestions: FieldCandidate[];
}

export interface AdvocateRowResult {
  row_index: number;
  status: FieldStatus;
  confidence: number;
  adv_code?: string | null;
  enrol_no?: string | null;
  enrol_year?: string | null;
  name?: string | null;
  mobile?: string | null;
  remark?: string | null;
  evidence: FieldEvidence[];
  suggestions: Record<string, FieldCandidate[]>;
}

export interface PartyMoreDetailsResult {
  status: FieldStatus;
  confidence: number;
  address?: string | null;
  district?: string | null;
  state?: string | null;
  pincode?: string | null;
  mobile?: string | null;
  email?: string | null;
  evidence: FieldEvidence[];
  suggestions: Record<string, FieldCandidate[]>;
}

export interface FilingGroupedResult {
  core_fields: Record<string, FieldResult>;
  petitioner_fields: Record<string, FieldResult>;
  respondent_fields: Record<string, FieldResult>;
  checkbox_fields: Record<string, FieldResult>;
  advocate_rows: AdvocateRowResult[];
  petitioner_more_details?: PartyMoreDetailsResult | null;
  respondent_more_details?: PartyMoreDetailsResult | null;
}

export interface ExtractionJobSummary {
  extraction_job_id: number;
  document_id: number;
  form_type: string;
  status: string;
  extractor_version: string;
  overall_confidence?: number | null;
  needs_review: boolean;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExtractionResponse {
  job: ExtractionJobSummary;
  fields: FieldResult[];
  grouped?: FilingGroupedResult | null;
  confirmed_count: number;
  suggested_count: number;
  missing_count: number;
  review_flags: string[];
}

export interface ExtractionRunRequest {
  triggered_by?: string | null;
  run_async: boolean;
  force_recompute: boolean;
  form_type: string;
}

File: backend/app/schemas/filing_extraction.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field


FieldStatus = Literal["confirmed", "suggested", "missing", "rejected"]
JobStatus = Literal["queued", "running", "completed", "needs_review", "failed"]


class FieldEvidence(BaseModel):
    source_type: str | None = None
    page_from: int | None = None
    page_to: int | None = None
    chunk_id: str | None = None
    text: str | None = None
    validation_notes: str | None = None


class FieldCandidate(BaseModel):
    value: str
    normalized_value: str | None = None
    confidence: float = 0.0
    status: FieldStatus = "suggested"
    evidence: FieldEvidence | None = None


class FieldResult(BaseModel):
    field_key: str
    field_label: str
    status: FieldStatus
    value: str | None = None
    normalized_value: str | None = None
    confidence: float = 0.0
    evidence: FieldEvidence | None = None
    suggestions: list[FieldCandidate] = Field(default_factory=list)


class AdvocateRowResult(BaseModel):
    row_index: int
    status: FieldStatus = "missing"
    confidence: float = 0.0
    adv_code: str | None = None
    enrol_no: str | None = None
    enrol_year: str | None = None
    name: str | None = None
    mobile: str | None = None
    remark: str | None = None
    evidence: list[FieldEvidence] = Field(default_factory=list)
    suggestions: dict[str, list[FieldCandidate]] = Field(default_factory=dict)


class PartyMoreDetailsResult(BaseModel):
    status: FieldStatus = "missing"
    confidence: float = 0.0
    address: str | None = None
    district: str | None = None
    state: str | None = None
    pincode: str | None = None
    mobile: str | None = None
    email: str | None = None
    evidence: list[FieldEvidence] = Field(default_factory=list)
    suggestions: dict[str, list[FieldCandidate]] = Field(default_factory=dict)


class FilingGroupedResult(BaseModel):
    core_fields: dict[str, FieldResult] = Field(default_factory=dict)
    petitioner_fields: dict[str, FieldResult] = Field(default_factory=dict)
    respondent_fields: dict[str, FieldResult] = Field(default_factory=dict)
    checkbox_fields: dict[str, FieldResult] = Field(default_factory=dict)
    advocate_rows: list[AdvocateRowResult] = Field(default_factory=list)
    petitioner_more_details: PartyMoreDetailsResult | None = None
    respondent_more_details: PartyMoreDetailsResult | None = None


class ExtractionRunRequest(BaseModel):
    triggered_by: str | None = None
    run_async: bool = True
    force_recompute: bool = False
    form_type: str = "filing_registration"


class ExtractionFeedbackItem(BaseModel):
    field_key: str
    system_value: str | None = None
    user_value: str
    correction_type: str = "edited"
    corrected_by: str | None = None


class ExtractionFeedbackRequest(BaseModel):
    items: list[ExtractionFeedbackItem]


class ExtractionJobSummary(BaseModel):
    extraction_job_id: int
    document_id: int
    form_type: str
    status: JobStatus
    extractor_version: str
    overall_confidence: float | None = None
    needs_review: bool = False
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ExtractionResponse(BaseModel):
    job: ExtractionJobSummary
    fields: list[FieldResult]
    grouped: FilingGroupedResult | None = None
    confirmed_count: int = 0
    suggested_count: int = 0
    missing_count: int = 0
    review_flags: list[str] = Field(default_factory=list)
